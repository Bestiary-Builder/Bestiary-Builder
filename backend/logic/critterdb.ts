import {app} from "../utilities/constants";
import {log} from "../utilities/logger";
import fetch from "node-fetch";

app.get("/api/critterdb/:id/:published", async (req, res) => {
	let id = req.params.id;
	let published = req.params.published.toLowerCase() == "true";
	let result = await from_critterdb(id, published);
	if (!result) {
		return res.status(500).json({error: "Failed to fetch info from critterdb.com. Are you sure the link is right?"});
	} else {
		return res.json(result);
	}
});

async function from_critterdb(url: string, published: boolean) {
	let data = {creatures: [] as any[], name: "", description: ""};
	log.info(`CritterDB | Getting bestiary ID ${url}...`);
	let api_base = published ? "https://critterdb.com:443/api/publishedbestiaries" : "https://critterdb.com:443/api/bestiaries";

	let errored = false;
	await fetch(`${api_base}/${url}`).then(async (resp) => {
		try {
			let raw = (await resp.json()) as any;
			data.name = raw["name"];
			data.description = raw["description"];
		} catch (error) {
			log.error(`CritterDB | Error importing bestiary metadata. Id: "${url}" Published: ${published}`);
			errored = true;
		}
	});
	if (!errored) {
		if (published) data.creatures = await get_published_bestiary_creatures(url, api_base);
		else data.creatures = (await get_link_shared_bestiary_creatures(url, api_base)) ?? [];
	}

	return errored ? null : data;
}

async function get_published_bestiary_creatures(id: string, api_base: string) {
	log.info(`CritterDB | Getting link published bestiary ${id}...`);
	let creatures = [] as any[];
	for (let index = 1; index <= 100 /* 100 pages max */; index++) {
		log.info(`CritterDB | Getting page ${index} of ${id}...`);
		let raw_creatures = (await fetch(`${api_base}/${id}/creatures/${index}`).then(async (resp) => {
			if (!(resp.status >= 200 && resp.status < 300)) {
				log.error(`CritterDB | Error importing published bestiary creatures. Id: "${id}".`);
				return null;
			}
			return resp.json();
		})) as any[] | null;
		if (!raw_creatures || raw_creatures.length == 0) break;
		creatures.push(raw_creatures);
	}
	return creatures.flat();
}

async function get_link_shared_bestiary_creatures(id: string, api_base: string) {
	log.info(`CritterDB | Getting link shared bestiary ${id}...`);
	let creatures = await fetch(`${api_base}/${id}/creatures`).then(async (resp) => {
		if (resp.status == 400) {
			log.error(`CritterDB | Permission error importing link shared bestiary creatures. Id: "${id}".`);
		} else if (!(resp.status >= 200 && resp.status < 300)) {
			log.error(`CritterDB | Unkown error importing link shared bestiary creatures. Id: "${id}".`);
			return null;
		}
		return (await resp.json()) as any[];
	});
	return creatures;
}
