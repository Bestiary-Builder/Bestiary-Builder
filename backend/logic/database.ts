import {MongoClient, ServerApiVersion} from "mongodb";
import type {Db, ObjectId, Collection} from "mongodb";
import {app} from "../server";
// Create a MongoClient with a MongoClientOptions object to set the Stable API version
const client = new MongoClient(process.env.MongoDB!, {
	serverApi: {
		version: ServerApiVersion.v1,
		strict: true,
		deprecationErrors: true
	}
});
//Connect to database
let database = null as Db | null;
startConnection();
async function startConnection() {
	try {
		// Connect the client to the server	(optional starting in v4.7)
		await client.connect();
		// Send a ping to confirm a successful connection
		database = client.db("bestiarybuilder");
		collections.users = database.collection("Users");
		console.log(`Successfully connected to database: ${database.databaseName} and collection: ${collections.users.collectionName}`);
	} catch (e: any) {
		console.error(e);
		// Ensures that the client will close on error
		await client.close();
	}
}
//Collections
export class User {
	constructor(public id: string, public username: string, public avatar: string, public email: string, public verified: boolean, public banner_color: string, public global_name: string, public creatures: ObjectId[] = [], public _id?: ObjectId) {}
}
const collections: {users?: Collection} = {};

//DB functions
export async function getUser(id: string) {
	let user = (await collections.users?.findOne({id: id})) as User;
	return user;
}
export async function updateUser(data: {id: string; username: string; avatar: string; email: string; verified: boolean; banner_color: string; global_name: string}) {
	console.log("Updating user", data);
	if ((await getUser(data.id)) != null) {
		console.log("Updating user with id " + data.id);
		collections.users?.updateOne({id: data.id}, {$set: data});
	} else {
		console.log("Added new user to collection");
		let userData = data as User;
		userData.creatures = [];
		collections.users?.insertOne(userData);
	}
	return await getUser(data.id);
}
