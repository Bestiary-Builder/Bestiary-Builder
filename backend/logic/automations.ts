import {app, checkAutomationLimits} from "../utilities/constants";
import {log} from "../utilities/logger";
import {requireUser, possibleUser} from "./login";
import {getUser, getAutomation, collections, updateAutomation, deleteAutomation} from "../utilities/database";
import {Automation, stringToId} from "../../shared";
import {checkBadwords} from "../utilities/badwords";

//Get info
app.get("/api/automation/:id", requireUser, async (req, res) => {
	try {
		let _id = stringToId(req.params.id);
		if (!_id) return res.status(400).json({error: "Automation id not valid."});
		let automation = await getAutomation(_id);
		if (!automation) {
			return res.status(404).json({error: "No automation with that id found."});
		}
		let user = req.body.user;
		if (automation.owner == user?._id) {
			//Return automation
			log.info(`Retrieved automation with the id ${_id}`);
			return res.json(automation);
		} else {
			return res.status(401).json({error: "You don't have access to this automation."});
		}
	} catch (err) {
		log.log("critical", err);
		return res.status(500).json({error: "Unknown server error occured, please try again."});
	}
});
app.get("/api/my-automations", requireUser, async (req, res) => {
	try {
		let user = req.body.user;
		if (!user) return res.status(404).json({error: "Couldn't find user"});
		let allAutomations = (await collections.automations?.find({owner: user._id}).toArray()) ?? [];
		log.info(`Retrieved ${allAutomations.length} automations from the current user with the id ${user._id}`);
		return res.json(allAutomations);
	} catch (err) {
		log.log("critical", err);
		return res.status(500).json({error: "Unknown server error occured, please try again."});
	}
});

app.get("/api/my-automations/list", requireUser, async (req, res) => {
	try {
		let user = req.body.user;
		if (!user) return res.status(404).json({error: "Couldn't find user"});
		let allAutomations = (await collections.automations?.find({owner: user._id}).toArray()) ?? [];
		log.info(`Retrieved all automations in list form from the current user with the id ${req.params.userid}`);
		return res.json(
			allAutomations.map((a) => {
				a.name, a._id;
			}) ?? []
		);
	} catch (err) {
		log.log("critical", err);
		return res.status(500).json({error: "Unknown server error occured, please try again."});
	}
});

//Update info
app.post("/api/automation/:id/update", requireUser, async (req, res) => {
	try {
		//Get input
		let user = req.body.user;
		if (!user) return res.status(404).json({error: "Couldn't find current user."});
		let _id = stringToId(req.params.id);
		if (!_id) return res.status(400).json({error: "Invalid automation id."});
		if (!req.body.data) return res.status(400).json({error: "Automation data not found."});
		let data = {
			...({
				_id: _id,
				automation: null,
				name: "New automation",
				description: "",
				owner: user._id
			} as Automation),
			...(req.body.data as Partial<Automation>)
		} as Automation;
		data._id = _id;
		//Check limits
		let limitError = checkAutomationLimits(data);
		if (limitError) return res.status(400).json({error: limitError});
		//Remove bad words
		let nameError = checkBadwords(data.name);
		if (nameError) return res.status(400).json({error: "Automation name " + nameError});
		let descError = checkBadwords(data.description);
		if (descError) return res.status(400).json({error: "Automation description " + descError});
		//Update existing automation
		let automation = await getAutomation(data._id);
		if (automation) {
			if (automation.owner != user._id) return res.status(401).json({error: "You don't have permission to update this automation."});
			//Limit properties that are editable:
			let update = {
				name: data.name,
				automation: data.automation,
				description: data.description
			};

			//Update:
			let updatedId = await updateAutomation(update as Automation, data._id);
			if (updatedId) {
				log.info(`Updated automation with the id ${data._id}`);
				return res.status(200).json(data);
			}
		} else {
			return res.status(404).json({error: "No automation with that id found."});
		}
	} catch (err) {
		log.log("critical", err);
		return res.status(500).json({error: "Unknown server error occured, please try again."});
	}
});

app.post("/api/automation/add", requireUser, async (req, res) => {
	try {
		//Get input
		let user = req.body.user;
		if (!user) return res.status(404).json({error: "Couldn't find current user."});
		if (!req.body.data) return res.status(400).json({error: "Automation data not found."});
		let data = {
			...({
				automation: null,
				name: "",
				description: "",
				owner: user._id
			} as Automation),
			...(req.body.data as Partial<Automation>)
		} as Automation;
		//Check limits
		let limitError = checkAutomationLimits(data);
		if (limitError) return res.status(400).json({error: limitError});
		//Remove bad words
		let nameError = checkBadwords(data.name);
		if (nameError) return res.status(400).json({error: "Automation name " + nameError});
		let descError = checkBadwords(data.description);
		if (descError) return res.status(400).json({error: "Automation description " + descError});
		//Create new automation
		let _id = await updateAutomation(data);
		if (!_id) return res.status(500).json({error: "Failed to create automation."});
		data._id = _id;
		data.owner = user._id!;
		log.info(`Created new automation with the id ${_id}`);
		return res.status(201).json(data);
	} catch (err) {
		log.log("critical", err);
		return res.status(500).json({error: "Unknown server error occured, please try again."});
	}
});

app.get("/api/automation/:id/delete", requireUser, async (req, res) => {
	try {
		//Get input
		let _id = stringToId(req.params.id);
		if (!_id) return res.status(400).json({error: "Automation id not valid."});
		let user = req.body.user;
		if (!user) return res.status(404).json({error: "Couldn't find current user."});
		//Permissions
		let automation = await getAutomation(_id);
		if (!automation) return res.status(404).json({error: "Couldn't find automation."});
		if (automation.owner != user._id) return res.status(401).json({error: "You don't have permission to delete this automation."});
		//Remove from db
		let status = await deleteAutomation(_id);
		if (status) {
			log.info(`Deleted automation with the id ${_id}`);
			res.json({});
		} else {
			res.status(500).json({error: "Failed to delete creature."});
		}
	} catch (err) {
		log.log("critical", err);
		return res.status(500).json({error: "Unknown server error occured, please try again."});
	}
});
