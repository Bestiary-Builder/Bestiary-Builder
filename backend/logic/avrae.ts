import { app } from "@/utilities/constants";
import { log } from "@/utilities/logger";
import { collections, getBestiary, incrementBestiaryViewCount } from "@/utilities/database";
import { Id, type InnateSpellsEntity, SKILLS_BY_STAT, type SkillsEntity, type SpellCasting, type Stat, type Statblock, componentsString, crAsString, displaySpeedOrSenses, fullSpellAbilityName, hpCalc, nthSuffix, ppCalc, spellAttackBonus, spellDc, statCalc, stringToId } from "~/shared";

// Export data
app.get("/api/public/bestiary/:id", (req, res) => res.redirect(`/api/export/bestiary/${req.params.id}`));
app.get("/api/export/bestiary/:id", async (req, res) => {
	/////////////////////////////////////////////////////////////////////////////////////
	/// / STOP. EDITING THIS FUNCTION CAN BREAK AVRAE IMPORTS. TEST BEFORE CHANGING  ////
	///////////////////////////////////////////////////////////////////////////////////
	try {
		const _id = stringToId(req.params.id);
		if (!_id)
			return res.status(400).json({ error: "Bestiary id not valid." });
		const bestiary = await getBestiary(_id);
		if (!bestiary)
			return res.status(404).json({ error: "No bestiary with that id found." });
		if (bestiary.status === "private")
			return res.status(401).json({ error: "This bestiary is private" });
		// Increment view count
		incrementBestiaryViewCount(_id);
		// Get creatures
		const creatures = [];
		for (const creatureId of bestiary.creatures) {
			const creature = await collections.creatures?.findOne({ _id: new Id(creatureId) });
			if (!creature)
				continue;

			const creatureData = getCreatureData(creature.stats);
			creatures.push(creatureData);
		}
		// Return bestiary in specific format
		const data = {
			metadata: {
				name: bestiary.name,
				description: bestiary.description
			},
			creatures
		};
		log.info(`Export - Retrieved bestiary with the id ${_id}`);
		return res.json(data);
	}
	catch (err) {
		log.log("critical", err);
		return res.status(500).json({ error: "Unknown server error occured. Please contact the developers of Bestiary Builder, not Avrae." });
	}
});

// Statblock functions:

function knownSpells(data: SpellCasting) {
	const dailySpells = {
		1: [],
		2: [],
		3: []
	} as {
		[key: string]: unknown[];
	};

	for (const times in data.innateSpells.spellList) {
		if (times === "0")
			continue;
		for (const sp of data.innateSpells.spellList[times])
			dailySpells[times].push(sp.spell);
	}
	const output = {
		caster_spells: data.casterSpells.spellList.flat(),
		at_will: data.innateSpells.spellList[0].map((sp: InnateSpellsEntity) => sp.spell),
		daily_spells: dailySpells
	};
	return output;
}

function calcSkills(data: Statblock) {
	const skillData = data.abilities.skills as SkillsEntity[];
	const output = {} as { [key: string]: { value: number; prof?: number; bonus: number; adv: number | null } };
	for (const stat in SKILLS_BY_STAT) {
		for (const skill of SKILLS_BY_STAT[stat as Stat]) {
			const raw = skillData.find(a => a.skillName.replaceAll(" ", "").toLowerCase() === skill.toLowerCase());
			if (raw === undefined) {
				output[skill] = {
					value: statCalc(stat as Stat, data),
					prof: 0,
					bonus: 0,
					adv: null
				};
			}
			else {
				if (raw.override != null) {
					output[skill] = {
						value: raw.override,
						// set prof to 1, as skills do not display in avrae without prof = 1|2
						prof: 1,
						bonus: 0,
						adv: null
					};
				}
				else {
					const base = statCalc(stat as Stat, data);
					if (raw.isHalfProficient) {
						output[skill] = {
							value: base + Math.floor(data.core.proficiencyBonus / 2),
							prof: 0.5,
							bonus: 0,
							adv: null
						};
					}
					else if (raw.isProficient) {
						output[skill] = {
							value: base + data.core.proficiencyBonus,
							prof: 1,
							bonus: 0,
							adv: null
						};
					}
					else if (raw.isExpertise) {
						output[skill] = {
							value: base + data.core.proficiencyBonus * 2,
							prof: 2,
							bonus: 0,
							adv: null
						};
					}
					else {
						output[skill] = {
							value: base,
							prof: 0,
							bonus: 0,
							adv: null
						};
					}
				}
			}
		}
	}
	return output;
}

function slots(num: number) {
	if (num > 1)
		return `(${num} slots)`;
	return `(${num} slot)`;
}

export function getCreatureData(creature: Statblock) {
	// HP:
	const hpObject = creature.defenses.hp;
	const hp = hpCalc(creature);
	const hitdice = `${`${hpObject.numOfHitDie}d${hpObject.sizeOfHitDie}`} + ${hpObject.numOfHitDie * statCalc("con", creature)}`;

	// Spellcastin:
	const spellcastInnateObj = creature.spellcasting.innateSpells;
	const spellcastCasterObj = creature.spellcasting.casterSpells;
	const spellcasting = {
		caster_level: spellcastCasterObj.casterLevel || 0,
		slots: spellcastCasterObj.spellSlotList || {},
		known_spells: knownSpells(creature.spellcasting),
		caster_dc: spellDc(false, creature),
		caster_sab: spellAttackBonus(false, creature),
		caster_mod: statCalc(spellcastCasterObj.spellCastingAbilityOverride ?? spellcastCasterObj.spellCastingAbility ?? null, creature),
		innate_dc: spellDc(true, creature),
		innate_sab: spellAttackBonus(true, creature),
		innate_mod: statCalc(spellcastInnateObj.spellCastingAbility, creature)
	};

	// Saves/stats
	const saves = {} as { [key: string]: unknown };
	for (const key in creature.abilities.saves) {
		const saveData = creature.abilities.saves[key as Stat];

		const newKey
			= {
				str: "strengthSave",
				dex: "dexteritySave",
				con: "constitutionSave",
				wis: "wisdomSave",
				int: "intelligenceSave",
				cha: "charismaSave"
			}[key] ?? "";

		const override = saveData.override;
		let value = statCalc(key as Stat, creature);
		let prof = 0;
		if (override != null) {
			value = override;
			prof = 1;
		}
		else if (saveData.isProficient) {
			value += creature.core.proficiencyBonus;
			prof = 1;
		}
		saves[newKey] = {
			value,
			prof,
			adv: null,
			bonus: 0
		};
	}
	// Final data
	const creatureData = {
		name: creature.description.name,
		proper: creature.description.isProperNoun,
		image_url: creature.description.image || "",
		languages: creature.core.languages,
		cr: crAsString(creature.description.cr),
		xp: creature.description.xp,
		alignment: creature.description.alignment,
		size: creature.core.size,
		race: creature.core.race,
		ac: creature.defenses.ac.ac,
		armortype: creature.defenses.ac.acSource,
		hp,
		hitdice,
		speed: displaySpeedOrSenses(creature.core.speed),
		ability_scores: {
			prof_bonus: creature.core.proficiencyBonus,
			strength: creature.abilities.stats.str,
			dexterity: creature.abilities.stats.dex,
			constitution: creature.abilities.stats.con,
			intelligence: creature.abilities.stats.int,
			wisdom: creature.abilities.stats.wis,
			charisma: creature.abilities.stats.cha
		},
		saves,
		skills: calcSkills(creature),
		senses: displaySpeedOrSenses(creature.core.senses),
		resistances: creature.defenses.resistances,
		immunities: creature.defenses.immunities,
		vulnerabilities: creature.defenses.vulnerabilities,
		condition_immune: creature.defenses.conditionImmunities,
		traits: creature.features.features,
		actions: creature.features.actions,
		bonus_actions: creature.features.bonus,
		reactions: creature.features.reactions,
		legactions: creature.features.legendary,
		mythic: creature.features.mythic,
		lair: creature.features.lair,
		regional: creature.features.regional,
		la_per_round: creature.misc.legActionsPerRound,
		spellcasting,
		passiveperc: ppCalc(creature)
	};

	const caster = creature.spellcasting.casterSpells;
	const isNoun = creature.description.isProperNoun;
	const name = creature.description.name;

	// best not to think about this too much.
	if (caster.casterLevel && caster.castingClass && caster.spellList.flat().length > 0) {
		const output = `${isNoun ? "" : "The "}${name} is a ${nthSuffix(caster.casterLevel)}-level spellcaster. ${isNoun ? "Their" : "Its"} spellcasting ability is ${fullSpellAbilityName(caster.spellCastingAbilityOverride ?? caster.spellCastingAbility)} (spell save DC ${spellDc(false, creature)}, ${
			spellAttackBonus(false, creature) >= 0 ? "+" : ""
		}${spellAttackBonus(false, creature)} to hit with spell attacks). ${isNoun ? name : "It"} ${
			["Sorcerer", "Bard", "Ranger", "Warlock"].includes(caster.castingClass) ? `knowns the following ${caster.castingClass.toLowerCase()} spells` : `has the following ${caster.castingClass.toLowerCase()} spells prepared`
		}:${!["Ranger", "Paladin"].includes(caster.castingClass) && caster.spellList[0].length > 0 ? `\n\nCantrips (at will): ${caster.spellList[0].sort().join(", ").toLowerCase()}` : ""}${
			caster.spellList[1].length > 0 ? `\n\n1st level ${slots(caster.spellSlotList![1])}: ${caster.spellList[1].sort().join(", ").toLowerCase()}` : ""
		}${caster.spellList[2].length > 0 ? `\n\n2nd level ${slots(caster.spellSlotList![2])}: ${caster.spellList[2].sort().join(", ").toLowerCase()}` : ""}${
			caster.spellList[3].length > 0 ? `\n\n3rd level ${slots(caster.spellSlotList![3])}: ${caster.spellList[3].sort().join(", ").toLowerCase()}` : ""
		}${caster.spellList[4].length > 0 ? `\n\n4th level ${slots(caster.spellSlotList![4])}: ${caster.spellList[4].sort().join(", ").toLowerCase()}` : ""}${
			caster.spellList[5].length > 0 ? `\n\n5th level ${slots(caster.spellSlotList![5])}: ${caster.spellList[5].sort().join(", ").toLowerCase()}` : ""
		}${caster.spellList[6].length > 0 ? `\n\n6th level ${slots(caster.spellSlotList![6])}: ${caster.spellList[6].sort().join(", ").toLowerCase()}` : ""}${
			caster.spellList[7].length > 0 ? `\n\n7th level ${slots(caster.spellSlotList![7])}: ${caster.spellList[7].sort().join(", ").toLowerCase()}` : ""
		}${caster.spellList[8].length > 0 ? `\n\n8th level ${slots(caster.spellSlotList![8])}: ${caster.spellList[8].sort().join(", ").toLowerCase()}` : ""}${
			caster.spellList[9].length > 0 ? `\n\n9th level ${slots(caster.spellSlotList![9])}: ${caster.spellList[9].sort().join(", ").toLowerCase()}` : ""
		}`.replaceAll("\t", "");

		creatureData.traits.push({
			name: "Spellcasting",
			description: output,
			automation: null
		});
	}

	const innateCaster = creature.spellcasting.innateSpells;
	if (innateCaster.spellCastingAbility && (innateCaster.spellList[0].length > 0 || innateCaster.spellList[1].length > 0 || innateCaster.spellList[2].length > 0 || innateCaster.spellList[3].length > 0)) {
		const fName = `Innate Spellcasting${innateCaster.isPsionics ? " (Psionics)" : ""}`;
		let output = "";

		if (!innateCaster.displayAsAction) {
			output += `${isNoun ? "" : "The "}${name}'s spellcasting ability is ${fullSpellAbilityName(innateCaster.spellCastingAbility)} (spell save DC ${spellDc(true, creature)}, ${spellAttackBonus(true, creature) >= 0 ? "+" : ""}${spellAttackBonus(true, creature)} to hit with spell attacks). ${
				isNoun ? name : "It"
			} can innately cast the following spells${componentsString(innateCaster.noComponentsOfType)}:`;
		}
		else {
			output += `${isNoun ? "" : "The "}${name} casts one of the following spells${componentsString(innateCaster.noComponentsOfType)} and using ${fullSpellAbilityName(innateCaster.spellCastingAbility)} as the spellcasting ability (spell save DC ${spellDc(true, creature)}, ${
				spellAttackBonus(true, creature) >= 0 ? "+" : ""
			}${spellAttackBonus(true, creature)} to hit with spell attacks):`;
		}

		if (innateCaster.spellList[0].length > 0) {
			output += `\n\nAt will: ${innateCaster.spellList[0]
				.map(x => (x.comment.length > 0 ? `${x.spell} (${x.comment})` : x.spell))
				.sort()
				.join(", ")
				.toLowerCase()}`;
		}
		if (innateCaster.spellList[3].length > 0) {
			output += `\n\n3/day each: ${innateCaster.spellList[3]
				.map(x => (x.comment.length > 0 ? `${x.spell} (${x.comment})` : x.spell))
				.sort()
				.join(", ")
				.toLowerCase()}`;
		}
		if (innateCaster.spellList[2].length > 0) {
			output += `\n\n2/day each: ${innateCaster.spellList[2]
				.map(x => (x.comment.length > 0 ? `${x.spell} (${x.comment})` : x.spell))
				.sort()
				.join(", ")
				.toLowerCase()}`;
		}
		if (innateCaster.spellList[1].length > 0) {
			output += `\n\n1/day each: ${innateCaster.spellList[1]
				.map(x => (x.comment.length > 0 ? `${x.spell} (${x.comment})` : x.spell))
				.sort()
				.join(", ")
				.toLowerCase()}`;
		}

		creatureData[innateCaster.displayAsAction ? "actions" : "traits"].push({
			name: fName,
			description: output,
			automation: null
		});
	}
	return creatureData;
}
