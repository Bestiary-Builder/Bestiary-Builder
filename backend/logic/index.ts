import "./bestiaries";
import "./creatures";
import "./critterdb";
import "./discord";
import "./login";
import "./search";
import "./staticData";
import "./user";
import "./validation";
import "./automations"