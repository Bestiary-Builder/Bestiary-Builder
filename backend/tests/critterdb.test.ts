import { expect, it } from 'vitest';
import { get_published_bestiary_creatures } from "../logic/critterdb";

it("CritterDB export to be unchanged", async () => {
    // Test https://critterdb.com/#/publishedbestiary/view/5acb0aa187653a455731b890 to be identical as before.
    const result = await get_published_bestiary_creatures("5acb0aa187653a455731b890", "https://critterdb.com:443/api/publishedbestiaries")
    expect(result).toMatchSnapshot();
})