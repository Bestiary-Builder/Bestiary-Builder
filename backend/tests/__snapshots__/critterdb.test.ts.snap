// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CritterDB export to be unchanged 1`] = `
[
  {
    "__v": 0,
    "_id": "5df1637163de6f61312fa47a",
    "flavor": {
      "description": "",
      "environment": "",
      "faction": "",
      "imageUrl": "http://avrae.io/assets/img/AvraeSquare.jpg",
      "nameIsProper": true,
    },
    "name": "Avrae",
    "publishedBestiaryId": "5acb0aa187653a455731b890",
    "sharing": {
      "linkSharingEnabled": false,
    },
    "stats": {
      "abilityScores": {
        "charisma": 18,
        "constitution": 14,
        "dexterity": 10,
        "intelligence": 22,
        "strength": 5,
        "wisdom": 12,
      },
      "actions": [
        {
          "description": "<i>Melee Weapon Attack:</i> +3 to hit, reach 5 ft., one creature. <i>Hit:</i> -1 (1d4 + -3) piercing damage, plus 35 (10d6) force damage.
<avrae hidden>Bite|3|1d4-3[piercing]+10d6[force]</avrae>",
          "name": "Bite",
        },
        {
          "description": "Avrae does a thing.",
          "name": "Do a Thing",
        },
        {
          "description": "<i>Melee Weapon Attack:</i> +6 to hit, reach 5 ft., one creature. <i>Hit:</i> 3 (1d6) slashing damage, or 4 (1d8) slashing damage if used with two hands to make a melee attack, plus 5 (1d10) force damage.
<avrae hidden>One Handed|6|1d6[slashing]+1d10[force]</avrae><avrae hidden>Two Handed|6|1d8[slashing]+1d10[force]</avrae>",
          "name": "Versatile Melee",
        },
        {
          "description": "<i>Melee or Ranged Weapon Attack:</i> +3 to hit, reach 5 ft. or range 80/320 ft., one creature. <i>Hit:</i> 0 (1d6 + -3) bludgeoning damage in melee, or 0 (1d6 + -3) bludgeoning damage at range, or 1 (1d8 + -3) bludgeoning damage if used with two hands to make a melee attack, plus 35 (10d6) force damage.",
          "name": "Throw Dice",
        },
      ],
      "additionalAbilities": [
        {
          "description": "Avrae's innate spellcasting ability is Intelligence (spell save DC 20, +12 to hit with spell attacks). Avrae can innately cast the following spells, requiring no components:

At will: <i>power word kill</i>, <i>power word heal</i>, <i>wish</i>, <i>power word panic</i>",
          "name": "Innate Spellcasting",
        },
        {
          "description": "Avrae is a 20th-level spellcaster. Avrae's spellcasting ability is Intelligence (spell save DC 20, +12 to hit with spell attacks). Avrae has the following wizard spells prepared:

3rd level (3 slots): <i>fireball</i>",
          "name": "Spellcasting",
        },
      ],
      "alignment": "Lawful Neutral",
      "armorClass": 16,
      "armorType": "natural armor",
      "challengeRating": 30,
      "conditionImmunities": [
        "Charmed",
        "Unconscious",
        "Stunned",
        "Petrified",
        "Paralyzed",
        "Exhaustion",
        "Intoxicated",
        "Frightened",
      ],
      "damageImmunities": [
        "Psychic",
        "Acid",
      ],
      "damageResistances": [],
      "damageVulnerabilities": [],
      "experiencePoints": 155000,
      "hitDieSize": 4,
      "languages": [
        "Common",
        "Draconic",
        "Python",
      ],
      "legendaryActions": [],
      "legendaryActionsDescription": "",
      "legendaryActionsPerRound": 3,
      "numHitDie": 10,
      "proficiencyBonus": 6,
      "race": "Dragon",
      "reactions": [],
      "savingThrows": [
        {
          "ability": "intelligence",
          "proficient": false,
          "value": 18,
        },
        {
          "ability": "constitution",
          "proficient": true,
        },
      ],
      "senses": [
        "truesight 120 ft.",
      ],
      "size": "Tiny",
      "skills": [
        {
          "name": "Arcana",
          "proficient": false,
          "value": 18,
        },
        {
          "name": "Perception",
          "proficient": true,
        },
      ],
      "speed": "30 ft.",
    },
  },
  {
    "__v": 0,
    "_id": "5df1637163de6f61312fa47b",
    "flavor": {
      "description": "",
      "environment": "",
      "faction": "",
      "imageUrl": "",
      "nameIsProper": false,
    },
    "name": "Spellcasting Test",
    "publishedBestiaryId": "5acb0aa187653a455731b890",
    "sharing": {
      "linkSharingEnabled": false,
    },
    "stats": {
      "abilityScores": {
        "charisma": 16,
        "constitution": 10,
        "dexterity": 10,
        "intelligence": 12,
        "strength": 10,
        "wisdom": 14,
      },
      "actions": [],
      "additionalAbilities": [
        {
          "description": "The spellcasting test's innate spellcasting ability is Charisma (spell save DC 13, +5 to hit with spell attacks). It can innately cast the following spells, requiring no components:

At will: <i>fire bolt</i>
3/day: <i>faerie fire</i>",
          "name": "Innate Spellcasting",
        },
        {
          "description": "The spellcasting test is a 10th-level spellcaster. The spellcasting test's spellcasting ability is Wisdom (spell save DC 12, +4 to hit with spell attacks). The spellcasting test has the following cleric spells prepared:

1st level (4 slots): <i>inflict wounds</i>, <i>bane</i>",
          "name": "Spellcasting",
        },
      ],
      "alignment": "Unaligned",
      "armorClass": 10,
      "armorType": null,
      "challengeRating": 0.125,
      "conditionImmunities": [],
      "damageImmunities": [],
      "damageResistances": [],
      "damageVulnerabilities": [],
      "experiencePoints": 25,
      "hitDieSize": 8,
      "languages": [
        "Common",
      ],
      "legendaryActions": [],
      "legendaryActionsDescription": "",
      "legendaryActionsPerRound": 3,
      "numHitDie": 1,
      "proficiencyBonus": 2,
      "race": "Humanoid",
      "reactions": [],
      "savingThrows": [],
      "senses": [],
      "size": "Medium",
      "skills": [],
      "speed": "30 ft.",
    },
  },
]
`;
